R/art.con.internal.R

```diff
@@ -0,0 +1,320 @@
+### interaction.term.labels: quoted interaction term label (e.g. "a:b:c")
+### concat.interaction.variable: concatenation (of type name) of all variables in interaction.variables (e.g., abc)
+#' @importFrom stats terms
+#' @importFrom pryr substitute_q
+#' @importFrom stringr str_replace
+parse.art.con.string.formula = function(f.orig){
+  
+  # check if f.orig is a single string (as opposed to a vector of multiple strings)
+  if(!is.character(f.orig) || length(f.orig) != 1){
+    stop("Contrast must be string term of form \"X1:X2\")")
+  }
+  
+  # don't allow spaces. must be "a:b", otherwise hard to parse
+  f.orig.has.space = grepl(' ', f.orig, fixed=TRUE)
+  if(f.orig.has.space){
+    stop("Contrast term cannot have a space in it.")
+  }
+  
+  # e.g. "a:b:c" -> list("a", "b", "c")
+  variables.str = strsplit(f.orig, ":")[[1]]
+  # e.g. list("a", "b", "c") -> list (a, b, c)
+  variables = lapply(variables.str, function(term) as.name(term))
+  
+  term.labels = f.orig
+  
+  # ensure we have exactly one interaction and no other terms
+  if (length(term.labels) != 1) {
+    stop("Model must have exactly one interaction and no other terms (" ,
+         length(term.labels), " terms given)")
+  }
+  
+  # Setup return list
+  # interaction.term.label: string name for interaction term label (e.g. "a:b:c")
+  interaction.term.label = term.labels
+  # interaction.variables: list of fixed variables (e.g., list(a, b, c))
+  interaction.variables = variables
+  # concat.interaction.variable: string formula with all
+  # interaction variables concatenated (e.g., abc)
+  concat.interaction.variable = as.name(paste(interaction.variables, collapse=""))
+  
+  list(
+    interaction.term.label = interaction.term.label,
+    interaction.variables = interaction.variables,
+    concat.interaction.variable = concat.interaction.variable
+  )
+}
+
+### parses f.orig which is formula or term (f) passed to art.con or artlm.con
+### f.orig can be of form ~ a*b*c or "a:b:c"
+### if f.orig is of "formula" form (i.e., ~ a*b*c),
+### converts it to "string" form (i.e., "a:b:c")
+### returns: result of parse.art.con.string.formula when passed "string" version of f.orig
+### list with
+### interaction.variables: list of fixed variables (of type name) in interaction (e.g., list(a, b, c)).
+### interaction.term.labels: quoted interaction term label (e.g. "a:b:c")
+### concat.interaction.variable: concatenation (of type name) of all variables in interaction.variables (e.g., abc)
+#' @importFrom stats as.formula
+#' @importFrom plyr is.formula
+parse.art.con.formula = function(f.orig){
+  
+  # looking for ~ a*b*c
+  if(is.formula(f.orig)){
+    # make sure : not in original formula
+    f.orig.str = deparse(f.orig)
+    f.orig.has.colon = grepl(':', f.orig.str, fixed=TRUE)
+    
+    if(f.orig.has.colon){
+      stop("Formula cannot contain : Did you mean ", as.formula(str_replace(f.orig.str, ':', ' * ')), 
```

Comment:
You should be able to use `gsub` instead of `str_replace` here (and elsewhere) and then drop the `stringr` dependency